ipa-core/src/utils/arraychunks.rs

```diff
@@ -0,0 +1,73 @@
+use crate::{ff::Fp61BitPrime, secret_sharing::SharedValue};
+
+#[allow(dead_code)]
+pub trait ArrayChunkIterator: Iterator {
+    /// This function returns an iterator that yields arrays of size `L`.
+    /// When the amount of items in the iterator is not a multiple of `L`
+    /// the iterator fills the last array with zero elements.
+    fn chunk_array<const L: usize>(self) -> ArrayChunk<Self, L>
+    where
+        Self: Sized,
+    {
+        ArrayChunk { iter: self }
+    }
+}
+
+pub struct ArrayChunk<I, const L: usize> {
+    iter: I,
+}
+
+impl<I: Iterator<Item = Fp61BitPrime>> ArrayChunkIterator for I {}
+
+#[allow(clippy::while_let_on_iterator)]
+impl<I, const L: usize> Iterator for ArrayChunk<I, L>
+where
+    I: Iterator<Item = Fp61BitPrime>,
+{
+    type Item = [I::Item; L];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut array = [Fp61BitPrime::ZERO; L];
+        let mut counter = 0usize;
+        while let Some(element) = self.iter.next() {
+            array[counter] = element;
+            counter += 1;
+            if counter == L {
+                break;
+            }
+        }
```

Comment:
I wonder whether this could be:

```rust
let mut array = [Fp61BitPrime::ZERO; L];
for i in 0..L {
  array[i] = self.iter.next().unwrap_or(Fp61BitPrime::ZERO);
}
array
```

Or even as short as:

```rust
self.iter.chain(repeat(Fp61BitPrime::ZERO)).take(L).collect::<[I::Item; L]>()
```

(I'm not sure about the ability to collect into an array though.)

Reply:
That would be nice. I didn't find a better way of doing it. There are two main issues, I need to increment the state of the iterator. The other issue is that if there is no element left, I want to return None rather than an array with all zeros.

Reply:
Ah, yes.  The code ends up being a little more complex.  The latter fails due to `chain()` consuming the iterator.  The former though...

```rust
if let Some(v) = self.iter.next() {
  let mut array = [Fp61BitPrime::ZERO; L];
  array[0] = v;
  for i in 1..L {
    array[i] = self.iter.next().unwrap_or(Fp61BitPrime::ZERO);
  }
  Some(array)
} else {
  None
}
```